src/app/main/main.component.ts

```diff
@@ -289,48 +387,84 @@ export class MainComponent implements OnInit, OnDestroy {
   }
 
   // 'count' is used to track the number of request. Each request is 10MB.
-  getResults(count: number, projectId: string, inputPageToken: string, location: string, jobID: string): Promise<BigQueryResponse> {
+  getResults(count: number, projectID: string, inputPageToken: string, location: string, jobID: string): Promise<BigQueryResponse> {
     if (!inputPageToken || count >= MAX_PAGES) {
       // Force an update feature here since everything is done.
       this.rows = this.rows.slice(0);
       return;
     }
     count = count + 1;
-    return this.dataService.getResults(projectId, jobID, location, inputPageToken, this.columns, this.stats).then(({ rows, stats, pageToken }) => {
+    return this.dataService.getResults(projectID, jobID, location, inputPageToken, this.columns, this.stats).then(({ rows, stats, pageToken }) => {
       this.rows.push(...rows);
       this.stats = stats;
       this._changeDetectorRef.detectChanges();
-      return this.getResults(count, projectId, pageToken, location, jobID);
+      return this.getResults(count, projectID, pageToken, location, jobID);
     });
   }
 
+  convertToUserQuery(geovizQuery : string) : string {
+    if (!geovizQuery) return '';
+
+    var lines = geovizQuery.split('\n');
+    var userQueryStarted = false;
+    var userQuery = '';
+    lines.forEach((line) => {
+      if (line.includes(USER_QUERY_START_MARKER)) {
+	userQueryStarted = true;
+      } else if (line.includes(USER_QUERY_END_MARKER)) {
+	userQueryStarted = false;
+      } else {
+	if (userQueryStarted) {
+	  userQuery += line + '\n';
+	}
+      }
+    });
+
+    return userQuery.trim();
+  }
+
+  convertToGeovizQuery(userQuery : string, geoColumns: BigQueryColumn[], numCols : number) :  string {
+    const hasNonGeoColumns = geoColumns.length < numCols;
+    const nonGeoClause = hasNonGeoColumns
+      ? `* EXCEPT(${geoColumns.map((f) => `\`${f.name}\``).join(', ') }),`
+      : '';
+    return `SELECT
+  ${nonGeoClause}
+  ${ geoColumns.map((f) => `ST_AsGeoJson(\`${f.name}\`) as \`${f.name}\``).join(', ') }
+FROM (
+${USER_QUERY_START_MARKER}\n
+${userQuery.replace(/;\s*$/, '')}\n
+${USER_QUERY_END_MARKER}\n
+);`;              
+  }
+
   query() {
     if (this.pending) { return; }
     this.pending = true;
 
-    const { projectID, sql, location } = this.dataFormGroup.getRawValue();
-
     // We will save the query information to local store to be restored next
     // time that the app is launched.
-    this.saveDataToLocalStorage(projectID, sql, location);
+    const dataFormValues = this.dataFormGroup.getRawValue();
+    this.projectID = dataFormValues.projectID;
+    const sql = dataFormValues.sql;
+    this.location = dataFormValues.location;
+    this.saveDataToLocalStorage(this.projectID, sql, this.location);
+    
+    // Clear the existing sharing URL.
+    this.clearGeneratedSharingUrl();
+    this.newSharingIdRequired = true;
```

Comment:
Probably not in this PR, but what do you think of removing the localstorage code now that we have remote storage available? I think that would help to keep things simple.

Reply:
That makes sense.